include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,311 @@
+/**
+ * @file nsga2_impl.hpp
+ * @author Sayan Goswami
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more Information.
+ */
+
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+#define ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+
+#include "nsga2.hpp"
+#include <set>
+#include <vector>
+#include <map>
+
+namespace ens {
+
+inline NSGA2::NSGA2(const size_t populationSize,
+                    const size_t maxGenerations,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const double epsilon) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon)
+{ /* Nothing to do here. */ }
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType,
+         typename... CallbackTypes>
+std::vector<MatType> NSGA2::Optimize(MultiobjectiveFunctionType& objectives,
+                        MatType& iterate,
+                        CallbackTypes&&... callbacks)
+{
+  std::vector<arma::vec> calculatedObjectives;
+  calculatedObjectives.resize(populationSize);
+
+  std::vector<MatType> population;
+  std::vector<std::vector<size_t> > fronts;
+  std::vector<double> crowdingDistance;
+  std::vector<size_t> ranks;
```

Comment:
I think we could use an `arma::uvec` here as well?

Reply:
Intermediate population size is `2 * populationSize`. Should we pre-allocate it?

Reply:
IMO it makes the most sense to do that. Another option would be to use `reserve()` to prevent reallocation of memory of the vector during addition, but as long as we add a comment for the reason why we have made it that size, pre-allocation should be fine.